multi_region_fft_solver.jl

```diff
@@ -0,0 +1,330 @@
+    Equation ``(∇² + m) ϕ = b`` is sometimes called the "screened Poisson" equation
+    when ``m < 0``, or the Helmholtz equation when ``m > 0``.
+"""
+function solve!(ϕ, solver::MultiRegionFFTBasedPoissonSolver, b, m=0)
+    arch = architecture(solver)
+    λx, λy, λz = solver.eigenvalues
+
+    # Temporarily store the solution in ϕc
+    ϕc = solver.storage
+
+    # Transform b *in-place* to eigenfunction space (we have to encode the)
+    @apply_regionally regional_transform!(b, solver, solver.transforms.forward[1])
+    # If we didn't have unified memory the transpose would go in here!
+    @apply_regionally regional_transform!(b, solver, solver.transforms.forward[2:3])
+
+    # Solve the discrete screened Poisson equation (∇² + m) ϕ = b.
+    @. ϕc = - b / (λx + λy + λz - m)
+
+    # If m === 0, the "zeroth mode" at `i, j, k = 1, 1, 1` is undetermined;
+    # we set this to zero by default. Another slant on this "problem" is that
+    # λx[1, 1, 1] + λy[1, 1, 1] + λz[1, 1, 1] = 0, which yields ϕ[1, 1, 1] = Inf or NaN.
+    m === 0 && CUDA.@allowscalar ϕc[1, 1, 1] = 0
+
+    # Apply backward transforms in order
+    # Transform b *in-place* to eigenfunction space (we have to encode the)
+    @apply_regionally regional_transform!(b, solver, solver.transforms.backward[1])
+    # If we didn't have unified memory the transpose would go in here!
+    @apply_regionally regional_transform!(b, solver, solver.transforms.backward[2:3])
+
+    copy_event = launch!(arch, solver.grid, :xyz, copy_real_component!, ϕ, ϕc, dependencies=device_event(arch))
+    wait(device(arch), copy_event)
+
+    return ϕ
+end
+
+# Here we have to encode the grid transformation
+function regional_transform!(b, solver, direction) 
+    # Perform FFT in two direction
+    [transform!(b, solver.buffer) for transform! in direction]
+end
+
+@kernel function copy_real_component!(ϕ, ϕc)
+    i, j, k = @index(Global, NTuple)
+    @inbounds ϕ[i, j, k] = real(ϕc[i, j, k])
+end
+
+import Oceananigans.Architectures: architecture
+
+abstract type AbstractTransformDirection end
+
+struct MultiRegionForward <: AbstractTransformDirection end
+struct MultiRegionBackward <: AbstractTransformDirection end
+
+#####
+##### Normalization factors
+#####
+
+normalization_factor(arch, topo, direction, N) = 1
+
+"""
+    normalization_factor(::CPU, ::Bounded, ::Backward, N)
+
+`FFTW.REDFT01` needs to be normalized by 1/2N.
+See: http://www.fftw.org/fftw3_doc/1d-Real_002deven-DFTs-_0028DCTs_0029.html#g_t1d-Real_002deven-DFTs-_0028DCTs_0029
+"""
+normalization_factor(::CPU, ::Bounded, ::Backward, N) = 1 / 2N
+
+#####
+##### Twiddle factors
+#####
+
+twiddle_factors(arch, grid, dim) = nothing
+
+"""
+    twiddle_factors(arch::GPU, grid, dims)
+
+Twiddle factors are needed to perform DCTs on the GPU. See equations (19a) and (22) of [Makhoul80](@cite)
+for the forward and backward twiddle factors respectively.
+"""
+function twiddle_factors(arch::GPU, grid::MultiRegionGrid, dims)
```

Comment:
I think it'd be better to reuse this stuff rather than copy/pasting it, because its pretty critical and non-trivial